src/Sylius/Bundle/OrderBundle/NumberGenerator/SequentialOrderNumberGenerator.php

```diff
@@ -64,11 +76,30 @@ public function __construct(
      */
     public function generate(OrderInterface $order)
     {
-        $sequence = $this->getSequence();
-        
-        $number = $this->generateNumber($sequence->getIndex());
-        $sequence->incrementIndex();
-        
+        $number = null;
+        $counter = 0;
+        $done = false;
+        while (!$done) {
+            if (++$counter > 100) {
+                $done = true;
+            }
+
+            $sequence = $this->getSequence();
+            $sequence = $this->sequenceManager->merge($sequence);
+
+            try {
+                $this->sequenceManager->lock($sequence, LockMode::OPTIMISTIC, $sequence->getVersion());
+
+                $number = $this->generateNumber($sequence->getIndex());
+                $sequence->incrementIndex();
+
+                $this->sequenceManager->flush();
+                $done = true;
+            } catch (OptimisticLockException $exception) {
+                $this->resetSequenceManagerIfClosed();
+            }
+        }
+
         return $number;
```

Comment:
We should not allow to place order without number, we probably need to throw an exception here.